[14기 송지원] step1 돔 조작과 이벤트 핸들링으로 메뉴 관리하기

zwonlala opened the pull request:
## 🎯 step1 요구사항 - 돔 조작과 이벤트 핸들링으로 메뉴 관리하기

- [x] 에스프레소 메뉴에 새로운 메뉴를 확인 버튼 또는 엔터키 입력으로 추가한다.
  - [x] 메뉴가 추가되고 나면, input은 빈 값으로 초기화한다.
  - [x] 사용자 입력값이 빈 값이라면 추가되지 않는다.
- [x] 메뉴의 수정 버튼을 눌러 메뉴 이름 수정할 수 있다.
  - [x] 메뉴 수정시 브라우저에서 제공하는 `prompt` 인터페이스를 활용한다.
- [x] 메뉴 삭제 버튼을 이용하여 메뉴 삭제할 수 있다.
  - [x] 메뉴 삭제시 브라우저에서 제공하는 `confirm` 인터페이스를 활용한다.
- [x] 총 메뉴 갯수를 count하여 상단에 보여준다.
- 추가되는 메뉴의 아래 마크업은 `<ul id="espresso-menu-list" class="mt-3 pl-0"></ul>` 안에 삽입해야 한다.

```js
<li class="menu-list-item d-flex items-center py-2">
  <span class="w-100 pl-2 menu-name">${name}</span>
  <button
    type="button"
    class="bg-gray-50 text-gray-500 text-sm mr-1 menu-edit-button"
  >
    수정
  </button>
  <button
    type="button"
    class="bg-gray-50 text-gray-500 text-sm menu-remove-button"
  >
    삭제
  </button>
</li>
```

<br>
<br>

## 😸 step1 코드리뷰 확인 사항  

1. 코드 불필요한 주석은 없는가? 
2. 불필요한 “console.log()” 는 없는가? 
3. eol 개행을 하였는가? 
	- [파일 끝에 개행을 추가해야 하는 이유](https://coderifleman.tumblr.com/post/115464362564/%ED%8C%8C%EC%9D%BC-%EB%81%9D%EC%97%90-%EA%B0%9C%ED%96%89%EC%9D%84-%EC%B6%94%EA%B0%80%ED%95%B4%EC%95%BC-%ED%95%98%EB%8A%94-%EC%9D%B4%EC%9C%A0)
	- [파일마다 EOL(End Of Line)을 왜 넣어야 할까](https://avocado12-note.tistory.com/11)
4. var 사용을 지양하고 let, const를 적절히 사용하였는가? 
5. 불필요한 전역변수를 사용하지 않았는가? 
6. 변수명에 적절한 네이밍을 적용하였는가?
7. 중복코드를 최소화하였는가?
8. 하나의 함수가 너무 많은 일을 하고 있진 않은가?
9. UI 렌더링 코드와 비즈니스 로직 코드를 분리하였는가?
10. [부정 조건을 사용하지 않는가?](https://velog.io/@seob/JavaScript-Clean-Code-%EA%B0%80%EB%8F%85%EC%84%B1-%EB%86%92%EC%9D%B4%EA%B8%B0)
11. [Early return pattern을 적용하였는가?](https://uminoh.tistory.com/m/35)

<br>
<br>

## 👨‍💻 코드 설명

### [Menu.js](https://github.com/blackcoffee-study/moonbucks-menu/pull/249/files#diff-4234773afef3ddbbe250f0090ac80c5c16bdeaae471602d62dc654065462d8b7)

- 메뉴에 대한 프로퍼티 및 함수가 정의되어 있는 클래스.

- 현재는 "step1"에서 메뉴에 관하여 관리하고 있는 상태가 "메뉴 이름" 밖에 없어 해당 코드 및 개념만 존재하는 상태
  - 추후 "메뉴 종류" 및 "품절" 상태 등이 추가되면 관련 프로퍼티 및 코드 추가될 예정

- 기존 TS를 사용할때와 다르게 JS에서는 파라미터의 타입을 특정할 수 없어,   
모든 입력 파라미터에 대해 `! 연산자`, `typeof 연산자` 등을 통해 유효한 값인지 입력을 체크하고 로직실행되게 구현함

<br>

### [MenuStatus.js](https://github.com/blackcoffee-study/moonbucks-menu/pull/249/files#diff-6f8c1349e30d21195c7ca11c93662934c6a28615a3d79a5f489b051d52d9dbe1)

- 이번 과제의 핵심? 인 사용자의 입력을 "상태"로 관리하는데 관련된 로직을 묶어놓은 class
- 프로퍼티로 `Menu []`을 저장하고 있고, 
  - 초기화 : [reset()](https://github.com/blackcoffee-study/moonbucks-menu/pull/249/files#diff-6f8c1349e30d21195c7ca11c93662934c6a28615a3d79a5f489b051d52d9dbe1#t=32:~:text=%7D-,reset,-()%20%7B)
  - 메뉴 추가 : [add(menu)](https://github.com/blackcoffee-study/moonbucks-menu/pull/249/files#diff-6f8c1349e30d21195c7ca11c93662934c6a28615a3d79a5f489b051d52d9dbe1#t=32:~:text=%7D-,add(newMenu),-%7B)
  - 메뉴 목록 조회 : [getMenuList()](https://github.com/blackcoffee-study/moonbucks-menu/pull/249/files#diff-6f8c1349e30d21195c7ca11c93662934c6a28615a3d79a5f489b051d52d9dbe1#t=32:~:text=%7D-,getMenuList(),-%7B)
  - 메뉴 개수 조회 : [getMenuCount()](https://github.com/blackcoffee-study/moonbucks-menu/pull/249/files#diff-6f8c1349e30d21195c7ca11c93662934c6a28615a3d79a5f489b051d52d9dbe1#t=32:~:text=%7D-,getMenuCount(),-%7B)
  - 메뉴 수정 : [update(index, newMenu)](https://github.com/blackcoffee-study/moonbucks-menu/pull/249/files#diff-6f8c1349e30d21195c7ca11c93662934c6a28615a3d79a5f489b051d52d9dbe1#t=32:~:text=update(index%2C%20newMenu))
  - 메뉴 삭제 : [delete(index)](https://github.com/blackcoffee-study/moonbucks-menu/pull/249/files#diff-6f8c1349e30d21195c7ca11c93662934c6a28615a3d79a5f489b051d52d9dbe1#t=32:~:text=*/-,delete(index),-%7B)
  - 메소드를 통해 상태를 업데이트 해줌
<br>

### [index.js](https://github.com/blackcoffee-study/moonbucks-menu/pull/249/files#diff-f168a78f0ef4dac15c2e638d98a4cfb4970cf18205cb7cd10f2cd3a58323c828)

UI 관련 로직과 함수가 정의되어 있는 파일

<br>
<br>

## 🤔 고민되는 부분

#### 0. 현재 구조가 새로운 메뉴가 추가되거나, 삭제되면 모든 메뉴들을 삭제한 다음, 다시 새로 메뉴들을 그려주는 구조임..
-> 퍼포먼스 적으로 좋지 않을 수도 있을 것 같은데 어떻게 개선하면 좋을까...? 🤔 🤔 🤔 

<br>

#### 1. [index.js](https://github.com/blackcoffee-study/moonbucks-menu/pull/249/files#diff-f168a78f0ef4dac15c2e638d98a4cfb4970cf18205cb7cd10f2cd3a58323c828) 파일 상단에 상태 값을 

```
const state = new MenuStatus([]);
```

이렇게 정의한 후, 각각의 메소드에서 위 state 값을 참조하여 로직을 실행

-> 뭔가 전역 변수를 선언하고 사용하는 느낌이라 문제되는 부분은 없을지 더 나은 구조는 무엇일지 고민이 됨 🤔 

<br>

#### 2. 위와 비슷하게 현재 CSS selector를 통해 `document.getElementById()` 함수 등으로

HTMLElement을 index.js 파일 상단에 정의하고 모든 코드에서 위 HTMLElement  참조하는 구조

-> 더 나은 구조는 없을까? 🤔 

<br>

#### 3. JS에서 함수 파라미터 입력이 valid 한지 검사하는 로직 부분에서 더 나은 로직이나 방법은 없을지? 🤔 

<br>

고민이 됩니다!



<br>
<br>

리뷰해주시면 감사하겠습니다!! 🙏

zwonlala committed:
docs: Add "춘식이는 개발개발해" team member

zwonlala committed:
docs: README.md 파일에 step 1 코드리뷰 확인사항 작성

zwonlala committed:
feat: Menu class 정의

사용자가 입력한 메뉴 항목에 대한 모델을 정의한 Menu class 정의

zwonlala committed:
feat: MenuStatus class 정의

메뉴판의 상태를 관리하는 클래스인 MenuStatus 클래스 정의

메뉴 초기화, 메뉴 추가, 메뉴 정보 불러오기, 메뉴 개수 리턴, 메뉴 수정, 메뉴 삭제하는 기능 존재

각 함수 파라미터 typeof, instanceof 명령어를 통해 유효한 입력인지 파악 후 함수 로직 수행하게 예외처리 로직 추가

zwonlala committed:
feat: Menu 클래스의 각 함수 파라미터의 유효성 검사 로직 추가

생성자 및 메뉴 이름 업데이트 하는 로직 실행 전
! 연산자 및 typeof 연산자로 유효한 string 값 입력되었는지 확인

zwonlala committed:
feat: index.html에 Menu.js, MenuStatus.js 파일 <script> 태그로 삽입

index.js 파일에서 Menu, MenuStatus 모듈 사용하기 위해
index.html 파일에 <script> 태그로 삽입

zwonlala committed:
feat: Input에 입력 후 엔터 키 입력하거나, 확인 버튼 클릭하면, MenuStatus 에 새로운 메뉴 추가되도록 구현

zwonlala committed:
feat: 메뉴 새로 추가되고 나면 새로운 메뉴 <li> 생성하여 렌더하는 로직 추가

zwonlala committed:
fix: 한글 입력 시 마지막 글자 추가적으로 입력되던 현상 수정

참고한 글 :
- https://velog.io/@corinthionia/JS-keydown%EC%97%90%EC%84%9C-%ED%95%9C%EA%B8%80-%EC%9E%85%EB%A0%A5-%EC%8B%9C-%EB%A7%88%EC%A7%80%EB%A7%89-%EC%9D%8C%EC%A0%88%EC%9D%B4-%EC%A4%91%EB%B3%B5-%EC%9E%85%EB%A0%A5%EB%90%98%EB%8A%94-%EA%B2%BD%EC%9A%B0-%ED%95%A8%EC%88%98%EA%B0%80-%EB%91%90-%EB%B2%88-%EC%8B%A4%ED%96%89%EB%90%98%EB%8A%94-%EA%B2%BD%EC%9A%B0
- https://levelup.gitconnected.com/javascript-events-handlers-keyboard-and-load-events-1b3e46a6b0c3

zwonlala committed:
chore: format index.js file

zwonlala committed:
fix: 사용자가 메뉴명 입력하지않고 Enter 입력하거나, 확인 버튼 누르는 케이스 예외처리

사용자가 메뉴명 입력하지않고 Enter 입력하거나, 확인 버튼 누르면
alert() 통해 추후 로직 진행되지 않게 수정

zwonlala committed:
feat: 메뉴 리스트 렌더링 시 총 메뉴 개수 업데이트 하는 로직 추기

zwonlala committed:
feat: 메뉴 명 업데이트 하는 로직 추가

zwonlala committed:
feat: 메뉴 아이템 삭제하는 로직 추가

zwonlala committed:
chore: EOF 개행 추가

zwonlala committed:
refactor: 변수명 i -> index 로 변경

zwonlala committed:
refactor: removeMenuItem() 함수 내 early return 하는 구조로 변경

zwonlala committed:
refactor: input 으로부터 사용자 입력 받아오는 중복 로직 getTrimmedInputString()로 함수화

zwonlala committed:
refactor: function 정의부 제외한 로직 파일 상단으로 위치 수정

hey210056 commented on src/js/index.js:
```diff
@@ -0,0 +1,110 @@
+	const inputString = getTrimmedInputString();
+
+	if (e.key === "Enter" && !e.isComposing) {
+		// 사용자가 <input>에 메뉴명 입력 없이 "Enter" 입력 시 예외 처리시키는 부분
```
isComposing 메서드에 대해 알아갑니다! 👍

hey210056 commented on src/js/index.js:
```diff
@@ -0,0 +1,110 @@
+		alert("값을 입력해주세요.");
+		return;
+	}
+	state.update(index, new Menu(newName.trim()));
```
trim으로 공백 제거하여 값을 처리하는 방법 좋네요!

hey210056 commented on src/js/index.js:
```diff
@@ -0,0 +1,110 @@
+	li.appendChild(menuName);
+	li.appendChild(editButton);
+	li.appendChild(deleteButton);
+
```
메이커준님 강의를 보다가 새롭게 배우게된 점이 있어 공유해드립니다!
mdn에 따르면 innerHTML보다는 insertAdjacentHTML의 사용을 권장합니다.
> 요소(element)의 내용을 변경하는 대신 HTML을 문서(document)에 삽입하려면, [insertAdjacentHTML()](https://developer.mozilla.org/ko/docs/Web/API/Element/insertAdjacentHTML) 메서드를 사용하십시오. 
[https://developer.mozilla.org/ko/docs/Web/API/Element/innerHTML](url)

zwonlala replied:
옹 넵넵. 알려주셔서 감사합니다!

찾아보니 "appendChild" 보다 "insertAdjacentHTML" 함수가 성능적으로 더 나은것 같네요!

알려주셔서 감사합니다!

hey210056 reviewed (commented):
다른 사람의 코드를 리뷰하는게 처음이라 서툰 부분이 많습니다.😅

bingwer commented on src/js/index.js:
```diff
@@ -0,0 +1,110 @@
+const inputTag = document.getElementById("espresso-menu-name");
+const submitButton = document.getElementById("espresso-menu-submit-button");
+const menuList = document.getElementById("espresso-menu-list");
+const menuCount = document.getElementsByClassName("menu-count")[0];
```
하나만 존재하는 element에 대해서는 querySelector를 사용하는게 좋을것같은데, getElementsByClassName를 사용하신 다른 이유가 있으신지 궁금합니다.

zwonlala replied:
개인적으로 `getElementById()` 사용하는게 익숙해서, 
해당 "menu-count" class 로 HTMLElement 를 찾을 때 `getElementsByClassName()` 함수를 사용하였습니다.

> 하나만 존재하는 element에 대해서는 querySelector를 사용하는게 좋을것같은데

@bingwer 혹시 승효님께서 생각하시는 위 내용에 대한 이유를 알려주실 수 있나요?
_`querySelector()` 함수 통해서는 첫 엘레멘트만 리턴하기 때문에 "[0]" 와 같은 부분들이 불필요해서_ 
일까요?

bingwer replied:
@zwonlala 
`getElementsByClassName()`은 document 전체를 탐색하지만
`querySelector()`는 일치하는 첫번째 요소를 만나면 반환하고 탐색을 중단하기 때문에
여러 요소를 한번에 찾을게 아니라면 `querySelector()`를 사용하는게 좀더 효과적이라 생각했어요!

------ 참고한 자료
[getElementsByClassName](https://developer.mozilla.org/ko/docs/Web/API/Element/getElementsByClassName)
[querySelector](https://developer.mozilla.org/ko/docs/Web/API/Document/querySelector)

zwonlala replied:
아 그르네욥! 놓쳤던 부분 알려주셔서 감사합니다! :)

bingwer replied:
@zwonlala 제가 해당부분에 대해 좀 더 찾아보았는데 퍼포먼스 적으로는 `getElement*`가 훨씬 좋다고 나오네요. 😂`querySelector`는 깊이 우선 탐색이라 자식노드의 수의 영향을 크게 받을것같아요.

[참고자료](https://stackoverflow.com/questions/14377590/queryselector-and-queryselectorall-vs-getelementsbyclassname-and-getelementbyid)

bingwer commented on src/js/index.js:
```diff
@@ -0,0 +1,110 @@
+const menuCount = document.getElementsByClassName("menu-count")[0];
+
+inputTag.addEventListener("keydown", onInputKeyDown);
+submitButton.addEventListener("click", onSubmitButtonClicked);
```
하나의 프로세스(메뉴를 입력하고, 등록한다)에 두개의 이벤트로 관리하는것보다 submit event 하나로 묶어서 관리하는것이 좋을것 같다고 생각하는데, 지원님은 어떻게 생각하시나요?

zwonlala replied:
아 그러네요~

submit event 하나만으로

- \<input>에 엔터 입력
- submit \<button> 클릭

두 이벤트를 한번에 처리 가능하네요!

submit 이벤트 사용하는 구조로 변경해보겠습니다~

bingwer commented on src/js/index.js:
```diff
@@ -0,0 +1,110 @@
+}
+
+function addNewMenu(menuName) {
+	const newMenu = new Menu(menuName);
```
저는 단순히 menu state를 객체로만 관리했었는데, class로 관리하는것도 좋은 방법 같네요! 👍

bingwer commented on src/js/index.js:
```diff
@@ -0,0 +1,110 @@
+function renderMenuList() {
+	// 기존에 menuList의 innerHtml에 존재하던 <li> 모두 제거
+	while (menuList.firstChild) {
+		menuList.removeChild(menuList.firstChild);
```
단순 자식노드삭제라면 remove()를 이용하는것이 더 좋을것 같습니다!

참고링크 : https://blogpack.tistory.com/683

zwonlala replied:
오 넵 좋은 글 추천 감사합니다! 👍

bingwer commented on src/js/index.js:
```diff
@@ -0,0 +1,110 @@
+		(menu, index) => menuList.appendChild(createMenuListItemElement(menu, index))
+	);
+
+	menuCount.innerHTML = `총 ${state.getMenuCount()}개`;
```
단순 text 입력이라면, HTMLParser가 포함된 innerHTML보다는 innerText또는 textContent를 사용하는게 더좋을것같습니다!

zwonlala replied:
넵 innerText 사용하는 편이 더 좋을 것 같네요! :)

bingwer reviewed (commented):
컴포넌트 부분만 클래스로 만들어 관리할 생각할 생각을 했지, 메뉴 자체는 클래스로 만들어서 관리할 생각을 못했었는데 많이 배워갑니다!

zwonlala committed:
refactor: PR 피드백 반영. appendChild -> insertAdjacentElement 로 변경

zwonlala committed:
refactor: PR 피드백 반영. "keydown", "click" 이벤트 관리하는 구조에서 "submit" 이벤트로 관리하는 구조로 변경

zwonlala committed:
refactor: PR 피드백 반영. 자식 노드들 삭제하는 로직 removeChild() -> remove() 사용하는 구조로 변경

zwonlala committed:
refactor: PR 피드백 반영. innerHTML -> innerText 사용하는 구조로 변경

zwonlala committed:
refactor: PR 피드백 반영. getElementsByClassName -> querySelector 로 변경

zwonlala committed:
refactor: event delegation 사용하는 구조로 변경

zwonlala committed:
refactor: 메뉴 <li> 생성하는 방식 문자열 + insertAdjacentHTML 사용하는 구조로 변경

zwonlala committed:
refactor: 메뉴 render 하는 로직 분리하여 함수화 시킨후, 각 함수 호출하는 전제 render 함수 사용하는 구조로 변경

메뉴 관련 data render 하는 로직을

- 메뉴 리스트 아이템 render
- 메뉴 개수 render

하는 부분으로 나누고

각각의 부분을

- renderMenuList()
- renderMenuCount()

함수로 정의함.

그리고 두 함수를 순차적으로 호출시키는 함수인 renderMenu() 함수 정의 후
data 변경사항 발생 시 renderMenu() 함수 호출되게 수정

zwonlala committed:
refactor: 메뉴 이름 수정하는 로직 수정

메뉴 이름 수정하는 로직 실행시
이미 render 된 전체 list item remove() 후 그려주는 구조가 아닌,
이름 변경된 메뉴 list item만 update 되는 구조로 변경

zwonlala committed:
refactor: 메뉴 삭제하는 로직 수정

메뉴 삭제하는 로직 실행시
이미 render 된 전체 list item remove() 후 그려주는 구조가 아닌,
이름 변경된 메뉴 list item만 삭제하고
메뉴 개수 부분 ui 다시 render 하는 구조로 변경

zwonlala commented:
@imakerjun 메이커준 님 merge 요청 드립니다! :)

imakerjun commented:
@zwonlala merge해드렸습니다 😆
